server/src/main/java/org/elasticsearch/common/util/concurrent/EsExecutors.java

```diff
@@ -266,6 +267,27 @@ public static String threadName(final String nodeName, final String namePrefix)
         return "elasticsearch" + (nodeName.isEmpty() ? "" : "[") + nodeName + (nodeName.isEmpty() ? "" : "]") + "[" + namePrefix + "]";
     }
 
+    // to be used in assertions only.
+    public static boolean differentExecutors(Thread thread1, Thread thread2) {
+        assert thread1 != thread2 : "only call for different threads";
```

Comment:
Probably want to return `false` here rather than failing with an unhelpful message?

Reply:
This was sort of on purpose in that, the method should only be used in situations where you are clearly comparing different threads. It seems that if you use this for the same thread your assertion is not written correctly. Hence more of an assertion on the assertion. Not too tied to this though.

Reply:
Oh I see, yeah, we can't be completing on the waiting thread because that thread is waiting...

(deserves a comment to avoid redoing that thinking tho)